brainframe_qt/api_utils/streaming/stream_manager.py

```diff
@@ -1,139 +1,156 @@
+        """
+        self._paused_streams: List[int] = []
+        """Currently paused streams"""
 
-        :param stream_config: The stream to connect to
-        :param url: The URL to stream on
-        :return: A Stream object
+        self.stream_readers: Dict[int, SyncedStreamReader] = {}
+        """All StreamReaders currently instantiated, paused or unpaused"""
+
+        self._init_signals()
+
+    def _init_signals(self) -> None:
+        self.destroyed.connect(self.close)
+
+    def close(self) -> None:
+        """Request and wait for all streams to close"""
+        logging.info("Initiating StreamManager close")
+        stream_readers = self.stream_readers.values()
+        for stream_reader in stream_readers:
+            stream_reader.close()
+        for stream_reader in stream_readers:
+            stream_reader.wait_until_closed()
+        logging.info("StreamManager close finished")
+
+    def delete_stream(self, stream_id: int, timeout: int = 120) -> None:
+        """[blocking API] Delete a stream through the API and initiate the closing of
+        its corresponding StreamReader
         """
-        if not self.is_streaming(stream_config.id):
-            # pipeline will be None if not in the options
-            pipeline: str = stream_config.connection_options.get("pipeline")
-
-            latency = StreamReader.DEFAULT_LATENCY
-            if stream_config.connection_type in self.REHOSTED_VIDEO_TYPES:
-                latency = StreamReader.REHOSTED_LATENCY
-            gobject_init.start()
-
-            # Streams created with a premises are always proxied from that
-            # premises
-            proxied = stream_config.premises_id is not None
-
-            stream_reader = GstStreamReader(
-                url,
-                latency=latency,
-                runtime_options=stream_config.runtime_options,
-                pipeline_str=pipeline,
-                proxied=proxied)
-            synced_stream_reader = SyncedStreamReader(
-                stream_config.id,
-                stream_reader,
-                self._status_receiver)
-            self._stream_readers[stream_config.id] = synced_stream_reader
-
-        return self._stream_readers[stream_config.id]
-
-    def is_streaming(self, stream_id: int) -> bool:
-        """Checks if the the manager has a stream reader for the given stream
-        id.
-
-        :param stream_id: The stream ID to check
-        :return: True if the stream manager has a stream reader, false
-            otherwise
+        self.stop_streaming(stream_id)
+        api.delete_stream_configuration(stream_id, timeout=timeout)
+
+    def pause_streaming(self, stream_id) -> None:
+        stream_reader = self.stream_readers[stream_id]
+        stream_reader.pause_streaming()
+
+        if stream_id in self._running_streams:
+            self._running_streams.remove(stream_id)
+
+        self._paused_streams.insert(0, stream_id)
```

Comment:
What advantage would an `OrderedDict` provide over a regular dict?

Reply:
I figure since you'd be using the fact that it's ordered as the main point, you would want to take advantage of the methods it supports that relate to order (like `reversed()`, `move_to_end(...)`, or `popitem(last=True)`)

Reply:
ooh I was unaware of `move_to_end`